core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala

```diff
@@ -67,6 +67,25 @@ private[master] class FileSystemRecoveryModeFactory(conf: SparkConf, serializer:
   }
 }
 
+/**
+ * LeaderAgent in this case is a no-op. Since leader is forever leader as the actual
+ * recovery is made by restoring from RocksDB.
+ */
+private[master] class RocksDBRecoveryModeFactory(conf: SparkConf, serializer: Serializer)
+  extends StandaloneRecoveryModeFactory(conf, serializer) with Logging {
+
+  val recoveryDir = conf.get(RECOVERY_DIRECTORY)
```

Comment:
It looks better to have a separate directory from the FileSystemRecoveryModeFactory

Reply:
+1

Reply:
Thank you for review. @yaooqinn and @LuciferYang 

I tried according to your advices but it seems to make the configuration namespace a little weird because this existing configuration is a general one, `spark.deploy.recoveryDirectory`.

https://github.com/apache/spark/blob/1d80d80a56c418f841e282ad753fad6671c3baae/core/src/main/scala/org/apache/spark/internal/config/Deploy.scala#L54

If we introduce a new configuration for `RocksDBPersistenceEngine` like the following, the users will ask why we cannot use `spark.deploy.recoveryDirectory` for `ROCKSDB` mode. Here is the example. The AS-IS design aims to allow the users to switch a single configuration `recoveryMode`.

**BEFORE**
```
spark.deploy.recoveryMode=FILESYSTEM
spark.deploy.recoveryDirectory=/opt/master
```

**AFTER**
```
spark.deploy.recoveryMode=ROCKSDB
spark.deploy.recoveryDirectory=/opt/master
```

If we add a new configuration, it would be the following where `spark.deploy.recoveryDirectory` is no-op.
```
spark.deploy.recoveryMode=ROCKSDB
spark.deploy.recoveryDirectory=/opt/master
spark.deploy.recoveryRocksDBDirectory=/opt/db
```

In addition, we assume that the users clear the location properly when they changes the setting like the following. 
- `spark.deploy.recoverySerializer`
- `spark.deploy.recoveryCompressionCodec`

To sum up, if you guys don't mind, I'd like to use the existing general configuration name for this backend.

Reply:
No, it will be okay. For example, rocksdb content will be ignored because FILESYSTEM scans only the following `prefix` files. I guess RocksDB also do the same things for the FILESYSTEM's leftover.

https://github.com/apache/spark/blob/1d80d80a56c418f841e282ad753fad6671c3baae/core/src/main/scala/org/apache/spark/deploy/master/PersistenceEngine.scala#L56-L78

However, that kind of changes are not supposed to be there. For the following three configurations, the directory should be cleared.

- spark.deploy.recoveryMode
- spark.deploy.recoverySerializer
- spark.deploy.recoveryCompressionCodec

Reply:
The thing is, it's pretty common to see that old naming conventions just don't cut it when adding new features.

Technically, we have the `spark.deploy.zookeeper.dir` for ZK-mode although it's a znode. 

I don't have a strong opinion for a new configuration here as it's trivial.

Reply:
Thank you, @yaooqinn . Ya, that's true. ZookeeperPersistenceEngine has a separate configuration namespace for that. That's one of the reasons why I don't want to add a new namespace.